rohmu/atomic_open.py

```diff
@@ -0,0 +1,40 @@
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any, BinaryIO, cast, ContextManager, Generator, IO, Optional, TextIO, Union
+
+import os
+
+
+def atomic_open_for_writing_text(final_path: Path, encoding: Optional[str] = None) -> ContextManager[TextIO]:
+    return cast(ContextManager[TextIO], atomic_opener(final_path, encoding=encoding, mode="w"))
+
+
+def atomic_open_for_writing_bytes(final_path: Path) -> ContextManager[BinaryIO]:
+    return cast(ContextManager[BinaryIO], atomic_opener(final_path, mode="wb"))
+
+
+@contextmanager
+def atomic_opener(
+    final_path: Path, mode: str, encoding: Optional[str] = None, _inode_spy=lambda inode: None
+) -> Generator[IO[Any], None, None]:
+    """
+    Creates a file object for writing which will only appear on the filesystem if the context manager succeeds.
+
+    It's designed to prevent partial writes (file will either be available with full content, or will be unavailable)
+    and the need for manual cleanup.
+    """
+    parent_dir = final_path.parent
+    if not parent_dir.exists():
+        raise IOError(f"Parent directory '{parent_dir}' must exist but is missing")
+    if "w" not in mode:
+        raise ValueError("Write mode must be used to make actual sense")
+
+    fd = os.open(str(parent_dir), os.O_TMPFILE | os.O_RDWR, 0o600)
```

Comment:
well, even if it did (but it throws an exception in Python nevertheless AFAICT) when using O_TMPFILE the semantics say that the file is deleted as soon as no fd references it. If it fails, no fd will reference this, and the file will be deleted.

Reply:
It throws an exception, but outside of the finally.
I doubt "as soon as no fd will reference this" applies to the `fd` var, since it's an int, so Python won't close the file descriptor when fd-the-var leaves the scope).
And I don't think this applies to "the abstract thing pointed to by fd", since it would allow linux/python to close the file descriptor even before `fdopen` is executed over the file descriptor. That abstract thing has no variable holding onto it.
So that would mean this applies to the end of the program, which means progressive leaking of fd in long running programs.

Reply:
Mmmh, maybe github is highlighting the wrong lines so my previous answer is unrelated.

Your concern is about os.fdopen failing when os.open has succeded, so we actually have an open file descriptor already.

I'd say that os.fdopen shouldn't throw, but we can't know, so yeah, we could add a check.

Reply:
Do you know a fun things?

os.fdopen may (if wrong encoding) or may not (if wrong mode) close the file descriptor which is passed in :-/ so we need to check what happens.